### PR TITLE
Use React Query instead of Redux

### LINKED ISSUE #100: Use React Query instead of Redux

### PR DESCRIPTION
**Pull request by mbehlendorf**
### Changed
* Use React Query instead of Redux to manage server-state #100
* Use TypeScript instead of Flow for type checking

resolves #100

### CODE REVIEW

**Review by sdorra — CHANGES_REQUESTED** [review]
Functionally the change looks good, but there are some things we should discuss.

This change contains a lot of mixed styles:

* Single Quotes vs Double Quotes
* Line end with semicolon and without
* Function definition with function keywork and as const
* isError vs error

We should consider using some tooling to force a consistent style. Eslint and prettier for example.

The other thing is the complete lack of types, even if we do not switch to typescript we should define types as much as we can. Each type helps the next developer to reach his goal faster and safer.

There is also a package-lock.json and a yarn.lock i’m not sure if the project uses npm or yarn.

File: src/main/js/App.js [hunk 1]
```diff
@@ -1,18 +1,25 @@
-import React, { Component } from 'react';
+import React, {Component} from 'react';
 import Navigation from './Navigation';
 import Main from './Main';
 import 'ces-theme/dist/css/ces.css';
 import {withRouter} from 'react-router-dom';
+import {QueryClient, QueryClientProvider} from 'react-query'
+import {ReactQueryDevtools} from 'react-query/devtools'
+
+const queryClient = new QueryClient({defaultOptions: {queries: {retry: 0}}})
```
[the remark this answers is not in the document]
Comment (mbehlendorf) [reply src/main/js/App.js]: done

File: src/main/js/App.js [hunk 2]
```diff
@@ -1,18 +1,25 @@
...
 
 class App extends Component {
```
[the remark this answers is not in the document]
Comment (mbehlendorf) [reply src/main/js/App.js:11]: done

File: src/main/js/repository/modules/repositories.js [hunk 3]
```diff
@@ -1,80 +1,7 @@
...
-            return {
-                ...state,
-                loading: false,
-                error: action.payload
-            };
-
-        default:
-            return state
-    }
+export function useRepositories() {
```
[the remark this answers is not in the document]
Comment (mbehlendorf) [reply src/main/js/repository/modules/repositories.js:4]: done

File: src/main/js/repository/modules/repository.js [hunk 4]
```diff
@@ -1,96 +1,7 @@
...
-                [action.id]: {
-                    loading: false,
-                    error: action.payload
-                }
-            };
-
-        default:
-            return state
-    }
+export function useRepository(id: string) {
```
[the remark this answers is not in the document]
Comment (mbehlendorf) [reply src/main/js/repository/modules/repository.js:4]: done

File: src/main/js/wiki/components/PageViewer.js [hunk 5]
```diff
@@ -9,24 +9,24 @@ type Props = {
...
     pagesLink: string,
     historyLink: string,
-    onDelete: () => void,
-    onHome: () => void,
-    onMove: () => void,
-    onRestore: () => void,
+    onDelete: any,
+    onHome: any,
+    onMove: any,
+    onRestore: any,
```
[the remark this answers is not in the document]
Comment (mbehlendorf) [reply src/main/js/wiki/components/PageViewer.js]: done

File: src/main/js/wiki/containers/Directory.js [hunk 6]
```diff
@@ -1,144 +1,92 @@
...
-    directory: any,
-    repository: string,
-    branch: string,
-    path: string,
-    url: string,
     t: any,
-    fetchWikiIfNeeded: (repository: string, branch: string) => void,
-    fetchDirectoryIfNeeded: (url: string) => void
+    match: any,
+    location: any
```
[the remark this answers is not in the document]
Comment (mbehlendorf) [reply src/main/js/wiki/containers/Directory.js]: done

File: src/main/js/wiki/containers/Directory.js [hunk 7]
```diff
@@ -1,144 +1,92 @@
...
-            return (
-                <div>
-                    <h1>Smeagol</h1>
-                </div>
-            );
-        }
+    const {repository, branch} = props.match.params;
 
+    const path = findDirectoryPath(props);
+    const {isLoading, error, data} = useDirectory(repository, branch, path)
```
[the remark this answers is not in the document]
Comment (mbehlendorf) [reply src/main/js/wiki/containers/Directory.js]: done

File: src/main/js/wiki/containers/History.js [hunk 8]
```diff
@@ -1,107 +1,63 @@
...
-    repository: string,
-    branch: string,
-    page: string,
-    url: string,
     t: any,
-    pagehistory: any,
-    fetchWikiIfNeeded: (repository: string, branch: string) => void,
-    fetchHistoryIfNeeded: (url: string) => void
+    match: any,
+    location: any
```
[the remark this answers is not in the document]
Comment (mbehlendorf) [reply src/main/js/wiki/containers/History.js]: done

File: src/main/js/wiki/containers/Page.js [hunk 9]
```diff
@@ -1,143 +1,139 @@
...
-                <div>
-                    <h1>Smeagol</h1>
-                </div>
-            );
-        }
-
-        if (editMode) {
-            return <PageEditor path={page.path} content={page.content} onSave={this.edit} onAbort={this.onAbortEdit} />;
-        }
+    const {repository, branch} = props.match.params;
```
[the remark this answers is not in the document]
Comment (mbehlendorf) [reply src/main/js/wiki/containers/Page.js]: done

File: src/main/js/wiki/containers/Page.js [hunk 10]
```diff
@@ -1,143 +1,139 @@
...
+    const path = findPagePath(props);
+
+    const pageQuery = usePage(repository, branch, path, getCommitParameter(props))
+    const wikiQuery = useWiki(repository, branch)
+
+    const editPageMutation = useEditPage(repository, branch, path)
+    const deletePageMutation = useDeletePage(repository, branch, path, pushLandingPageState)
+    const createPageMutation = useCreatePage(repository, branch, path)
+    const renamePageMutation = useRenamePage(repository, branch, path, pushPageStateClosure())
+    const restorePageMutation = useRestorePage(repository, branch, path,  pushPageStateClosure())
```
[the remark this answers is not in the document]
Comment (mbehlendorf) [reply src/main/js/wiki/containers/Page.js]: done

File: src/main/js/wiki/containers/Page.js [hunk 11]
```diff
@@ -1,143 +1,139 @@
...
+            </div>
+        );
+    }
 
-        return <PageViewer page={page} wiki={wiki} onDelete={ this.delete } onHome={ this.pushLandingPageState }
-                           onMove={ this.onMove } pagesLink={pagesLink} historyLink={historyLink}
-                           onRestore={this.onRestore} search={this.search}/>;
+    let wiki = wikiQuery.data;
+    wiki.repository = repository;
+    wiki.branch = branch;
```
[the remark this answers is not in the document]
Comment (mbehlendorf) [reply src/main/js/wiki/containers/Page.js]: done

File: src/main/js/wiki/containers/Search.js [hunk 12]
```diff
@@ -1,118 +1,78 @@
...
 type Props = {
-    loading: boolean,
-    error: Error,
-    results: any,
-    url: string,
-    fetchSearchResultsIfNeeded: (string) => void,
-    wiki: any
+    match: any,
+    history: any,
+    location: any
```
[the remark this answers is not in the document]
Comment (mbehlendorf) [reply src/main/js/wiki/containers/Search.js]: done

File: src/main/js/wiki/containers/Search.js [hunk 13]
```diff
@@ -1,118 +1,78 @@
...
-    createPageLink = (path: string) => {
-        const { repository, branch } = this.props;
+    let createPageLink = (path: string) => {
+        const {repository, branch} = props.match.params;
         return `/${repository}/${branch}/${path}`;
     };
 
-    search = (query: string) => {
-        const { history } = this.props;
+    let search = (query: string) => {
```
[the remark this answers is not in the document]
Comment (mbehlendorf) [reply src/main/js/wiki/containers/Search.js]: done

File: src/main/js/wiki/containers/Search.js [hunk 14]
```diff
@@ -1,118 +1,78 @@
...
-        const homeLink = this.createHomeLink();
-        let results = this.props.results;
+    const {repository, branch} = props.match.params;
+    const query = getQuery(props);
 
-        if (!results) {
-            results = [];
-        }
+    const searchQuery = useSearch(repository, branch, query)
+    const wikiQuery = useWiki(repository, branch)
```
[the remark this answers is not in the document]
Comment (mbehlendorf) [reply src/main/js/wiki/containers/Search.js]: done

File: src/main/js/wiki/modules/directory.js [hunk 15]
```diff
@@ -1,97 +1,13 @@
...
-                ...state,
-                [action.url] : {
-                    loading: false,
-                    error: action.payload
-                }
-            };
-        default:
-            return state
-    }
+export function useDirectory(repository: string,branch:string,path:string) {
```
[the remark this answers is not in the document]
Comment (mbehlendorf) [reply src/main/js/wiki/modules/directory.js]: done

File: src/main/js/wiki/modules/page.js [hunk 16]
```diff
@@ -1,336 +1,106 @@
 //@flow
```
[the remark this answers is not in the document]
Comment (mbehlendorf) [reply src/main/js/wiki/modules/page.js:1]: done

File: src/main/js/wiki/modules/pagehistory.js [hunk 17]
```diff
@@ -1,118 +1,13 @@
 // @flow
```
[the remark this answers is not in the document]
Comment (mbehlendorf) [reply src/main/js/wiki/modules/pagehistory.js:1]: done

File: src/main/js/wiki/modules/search.js [hunk 18]
```diff
@@ -1,97 +1,12 @@
 import {apiClient} from "../../apiclient";
```
[the remark this answers is not in the document]
Comment (mbehlendorf) [reply src/main/js/wiki/modules/search.js]: done

File: src/main/js/wiki/modules/wiki.js [hunk 19]
```diff
@@ -1,127 +1,8 @@
 //@flow
```
[the remark this answers is not in the document]
Comment (mbehlendorf) [reply src/main/js/wiki/modules/wiki.js:1]: done

File: src/main/js/wiki/containers/Page.js [hunk 20]
```diff
@@ -1,143 +1,139 @@
...
-    pushPageStateClosure = (pagePath: string) => {
-        const { history, repository, branch } = this.props;
-        return function() {
-            history.push(`/${repository}/${branch}/${pagePath}`);
+    const pushPageStateClosure = () => {
+        const {repository, branch} = props.match.params;
+        return function (pagePath: string) {
+            props.history.push(`/${repository}/${branch}/${pagePath}`);
         }
     };
```
[the remark this answers is not in the document]
Comment (mbehlendorf) [reply src/main/js/wiki/containers/Page.js]: done

File: package.json [hunk 21]
```diff
@@ -20,52 +20,56 @@
...
+    "@babel/core": "^7.13.1",
+    "@babel/plugin-proposal-class-properties": "^7.13.0",
+    "@babel/plugin-proposal-optional-chaining": "^7.13.0",
+    "@babel/plugin-transform-runtime": "^7.13.7",
+    "@babel/preset-env": "^7.13.5",
+    "@babel/preset-react": "^7.12.13",
+    "@babel/preset-typescript": "^7.13.0",
+    "@types/react": "^17.0.2",
+    "@types/react-dom": "^17.0.1",
+    "@types/react-router": "^5.1.11",
```
Comment (sdorra) [inline package.json:39]: The version of the types package should match the one from the dependencies.

**Review by sdorra — CHANGES_REQUESTED** [review]

**Pull request merged**